File: flask/ebaydata.py
from flask import Flask,request, render_template
import requests

import os
from ebay_oauth_token import OAuthToken



api_key = "WilliamL-hw2-PRD-aaf878574-a94ce998"
# findItemsAdvanced
def get_data(keyword,minPrice,maxPrice,conditions,seller_return_accepted,free_shipping,expedited_shipping,sort_order='BestMatch'):
    ebay_api_url = 'https://svcs.ebay.com/services/search/FindingService/v1'

    ebay_api_params = {
        'OPERATION-NAME': 'findItemsAdvanced',
        'SERVICE-VERSION': '1.0.0',
        'SECURITY-APPNAME':api_key, 
        'RESPONSE-DATA-FORMAT': 'JSON',
        'REST-PAYLOAD': 'true',
        'keywords': keyword,
        'paginationInput.entriesPerPage': '100',  
        
    }

    itemCounter = 0 # itemFilter counter

    if maxPrice:
        ebay_api_params[f"itemFilter({itemCounter}).name"]="MaxPrice"
        ebay_api_params[f"itemFilter({itemCounter}).value"] = maxPrice
        ebay_api_params[f"itemFilter({itemCounter}).paramName"] = "Currency"
        ebay_api_params[f"itemFilter({itemCounter}).paramValue"] = "USD"
        itemCounter += 1

    if minPrice:
        ebay_api_params[f"itemFilter({itemCounter}).name"]="MinPrice"
        ebay_api_params[f"itemFilter({itemCounter}).value"] = minPrice
        ebay_api_params[f"itemFilter({itemCounter}).paramName"] = "Currency"
        ebay_api_params[f"itemFilter({itemCounter}).paramValue"] = "USD"
        itemCounter += 1

    
    if conditions:
        ebay_api_params[f'itemFilter({itemCounter}).name'] = 'Condition'
        for i, condition in enumerate(conditions):
            ebay_api_params[f'itemFilter({itemCounter}).value({i})'] = condition
        itemCounter += 1

    if seller_return_accepted:
        ebay_api_params[f"itemFilter({itemCounter}).name"] = "ReturnsAcceptedOnly"
        ebay_api_params[f"itemFilter({itemCounter}).value"] = "true" if seller_return_accepted else "false"
        itemCounter += 1

        

    if free_shipping:
        ebay_api_params[f"itemFilter({itemCounter}).name"] = "FreeShippingOnly"
        ebay_api_params[f"itemFilter({itemCounter}).value"] = "true"
        itemCounter += 1

    if expedited_shipping:
        ebay_api_params[f"itemFilter({itemCounter}).name"] = "ExpeditedShippingType"
        ebay_api_params[f"itemFilter({itemCounter}).value"] = "Expedited"
        itemCounter += 1

    
    
    if sort_order is not None:
        ebay_api_params["sortOrder"] = sort_order

    # make the dictionary into a string url
    query_string = '&'.join([f'{key}={value}' for key, value in ebay_api_params.items()])
    ebay_api_url= ebay_api_url+'?'+query_string

    # print('the url you generate:',ebay_api_url)
    response = requests.get(ebay_api_url).json()
    
    return response


# getSingleItem
client_id =api_key
client_secret = "PRD-af878574c693-c7c9-4e95-b263-9852"
oauth_utility = OAuthToken(client_id, client_secret)
application_token = oauth_utility.getApplicationToken()
def get_single_data(item_id):
    headers = {
    "X-EBAY-API-IAF-TOKEN": oauth_utility.getApplicationToken(),
    }
    
    response_single = requests.get(f'https://open.api.ebay.com/shopping?callname=GetSingleItem&responseencoding=JSON&appid=WilliamL-hw2-PRD-aaf878574-a94ce998&siteid=0&version=967&ItemID={item_id}&IncludeSelector=Description,Details,ItemSpecifics', headers=headers)
    
    return response_single.json()
# print(get_single_data('175901174512'))

File: nodejs-angular/ebaydata.js
const axios = require('axios');
const client_id = 'WilliamL-hw2-PRD-aaf878574-a94ce998';
const client_secret = 'PRD-af878574c693-c7c9-4e95-b263-9852';
const OAuthToken = require('./ebay_oauth_token.js');

const oauth_utility = new OAuthToken(client_id, client_secret);
const api_key = 'WilliamL-hw2-PRD-aaf878574-a94ce998';

async function get_data(keyword, category, condition, shippingOptions, distanceMiles, postal) {
    const ebayApiUrl = 'https://svcs.ebay.com/services/search/FindingService/v1';
    const categoryMappings = {
        'Art': 550,
        'Baby': 2984,
        'Books': 267,
        'Clothing, Shoes, Accessories': 11450,
        'Computers/Tablets & Networking': 58058,
        'Health & Beauty': 26395,
        'Music': 11233,
        'Video Games & Consoles': 1249
    };
    const ebayApiParams = {
        'OPERATION-NAME': 'findItemsAdvanced',
        'SERVICE-VERSION': '1.0.0',
        'SECURITY-APPNAME': api_key,
        'RESPONSE-DATA-FORMAT': 'JSON',
        'REST-PAYLOAD': 'true',
        'keywords': keyword,
        'paginationInput.entriesPerPage': '50',
    };

    let itemFilterIndex = 0;

    if (distanceMiles) {
        ebayApiParams[`itemFilter(${itemFilterIndex}).name`] = 'MaxDistance';
        ebayApiParams[`itemFilter(${itemFilterIndex}).value`] = distanceMiles;
        itemFilterIndex++;
    }
    let conditionIndex = 0;
    if (condition && condition.length > 0) {
        console.log('i go inside')
        const conditions = Array.isArray(condition) ? condition : [condition]; // Ensure it's an array
        ebayApiParams[`itemFilter(${itemFilterIndex}).name`] = 'Condition';
    
        conditions.forEach((value, index) => {
            ebayApiParams[`itemFilter(${itemFilterIndex}).value(${index})`] = value;
        });
    
        itemFilterIndex++;
    }
    

    if (shippingOptions && shippingOptions.length > 0) {
        if (shippingOptions.includes('Local Pickup')) {
            ebayApiParams[`itemFilter(${itemFilterIndex}).name`] = 'LocalPickupOnly';
            ebayApiParams[`itemFilter(${itemFilterIndex}).value`] = 'true';
            itemFilterIndex++;
        }

        if (shippingOptions.includes('Free Shipping')) {
            ebayApiParams[`itemFilter(${itemFilterIndex}).name`] = 'FreeShippingOnly';
            ebayApiParams[`itemFilter(${itemFilterIndex}).value`] = 'true';
            itemFilterIndex++;
        }
    }

    if (category) {
        
        for (const cat in categoryMappings) {
            if (cat.toLowerCase().includes(category.toLowerCase())) {
                ebayApiParams['categoryId'] = categoryMappings[cat];
                break;
            }
        }
    }

    if (postal) {
        ebayApiParams['buyerPostalCode'] = postal;
    }

    const query_params = Object.keys(ebayApiParams)
        .map(key => `${key}=${encodeURIComponent(ebayApiParams[key])}`)
        .join('&');

    const ebayApiUrlWithParams = `${ebayApiUrl}?${query_params}&outputSelector(0)=SellerInfo&outputSelector(1)=StoreInfo`;// need to add this to match the instruction

    console.log('Constructed URL:', ebayApiUrlWithParams);

    try {
        // Send a GET request to the constructed URL
        const response = await axios.get(ebayApiUrlWithParams);
        console.log('eBay API Response:', response.data);
        return response.data;
    } catch (error) {
        console.error(error);
        throw error;
    }
}

// Example usage:
// get_data('iphone', 'Computers/Tablets & Networking', ['New','Used'], ['Free Shipping', 'Local Pickup'], 10, '90007');



  // Usage example


// async function getApplicationToken() {
//     try {
//         return await oauth_utility.getApplicationToken();
//     } catch (error) {
//         console.error('Error obtaining application token:', error);
//         throw error;
//     }
// }

async function get_single_data(item_id) {
    const headers = {
        "X-EBAY-API-IAF-TOKEN": await oauth_utility.getApplicationToken(),
    };

    const url = `https://open.api.ebay.com/shopping?callname=GetSingleItem&responseencoding=JSON&appid=WilliamL-hw2-PRD-aaf878574-a94ce998&siteid=0&version=967&ItemID=${item_id}&IncludeSelector=Description,Details,ItemSpecifics`;

    try {
        const response = await axios.get(url, { headers });
        return response.data;
    } catch (error) {
        console.error('Error obtaining single data:', error);
        throw error;
    }
}
// module.exports = get_single_data;
module.exports = {
    get_data,
    get_single_data
  };